<a href="https://colab.research.google.com/github/komisarzGiT/gai/blob/main/GUNI_ML_assignment1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Name: Marcin Byra  
Class: 2526-1MAO2  
Student ID: 25258161

The objective is to present end to end process of machine learning model training and evaluation based on provided train and test data sets.  

ML aghoritms must include Decision Tree classifiers:
Gini  

*   Gini
*   Information Gain
*   Gain ratuio

as well as either KNN or Random Fortst. KNN was chosen for this report to demonstarte different approach, since random forest is based on decision tree method.

**Problem setup, data loading, and cleaning**  

THere are two files provided:

Identify features vs. target, note any missing values or obvious anomalies, and briefly justify any minimal preprocessing (cleaning/filling missing values, etc.,) you apply.

In [9]:
#import liblaries
# standard imports
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#install c45 iblary
!pip install -U c45-decision-tree #https://pypi.org/project/c45-decision-tree/


In [10]:
#two files are provided containing respecively train and test data sets
train_data = 'https://raw.githubusercontent.com/komisarzGiT/gai/refs/heads/main/GU_ML01/Assignment1/EEG_Train_Binary.csv'
test_data = 'https://raw.githubusercontent.com/komisarzGiT/gai/refs/heads/main/GU_ML01/Assignment1/EEG_Test_Binary.csv'

df_train = pd.read_csv(train_data)
df_test = pd.read_csv(test_data)

train_rows, train_columns = df_train.shape
print(f"Training data set consist of {train_rows} observations and {train_columns} variables. Sample data:")
display(df_train.head(2))

test_rows, test_columns = df_test.shape
print(f"\nTraining data set consist of {test_rows} observations and {test_columns} variables. Sample data:")
display(df_test.head(2))

#check for missing values and outliers
print(f"\nStatistics to check for missing values and outliers")
display(df_train.describe().round(2).T)



Training data set consist of 3171 observations and 17 variables. Sample data:


X1   X2   X3  X4  X5   X6   X7   X8   X9  X10  X11  X12  X13  X14  X15  \
0   41   32  -27 -51 -89 -117  -83  -35   25   33   27   33   -9  -38  -57   
1 -281 -234 -166 -93 -22   55  109  137  115   80   54   50   37   -3  -53   

   X16  y  
0  -87  0  
1 -111  0


Training data set consist of 829 observations and 17 variables. Sample data:


X1  X2  X3  X4  X5  X6  X7  X8  X9  X10  X11  X12  X13  X14  X15  X16  y
0   68  46  26  11  28  57  45  23  83   93   35   30   38   26    4   21  0
1  104  32  26  11  -8  -3  26  43  31   56   73   -1  -25  -35  -81 -137  0


Statistics to check for missing values and outliers


count  mean    std    min   25%  50%   75%    max
X1   3171.0 -6.06  49.55 -281.0 -14.5  0.0   8.0  244.0
X2   3171.0 -5.97  49.52 -255.0 -17.0 -1.0  10.0  261.0
X3   3171.0 -6.09  48.90 -255.0 -12.5 -1.0   7.0  238.0
X4   3171.0 -5.89  49.28 -257.0 -18.0 -2.0  12.0  246.0
X5   3171.0 -6.18  49.69 -262.0 -24.0 -1.0  13.0  249.0
X6   3171.0 -5.93  49.37 -277.0 -18.0 -1.0  10.0  229.0
X7   3171.0 -6.13  49.31 -277.0 -15.0 -1.0   9.0  205.0
X8   3171.0 -6.22  50.28 -260.0 -19.0 -2.0  11.0  201.0
X9   3171.0 -6.55  50.72 -290.0 -18.0 -1.0   8.0  225.0
X10  3171.0 -7.00  51.28 -302.0 -21.0 -1.0  10.0  227.0
X11  3171.0 -7.45  50.78 -266.0 -15.5 -1.0   7.0  262.0
X12  3171.0 -7.52  51.04 -266.0 -20.0 -2.0  10.0  283.0
X13  3171.0 -7.54  51.25 -254.0 -24.0 -1.0  12.0  296.0
X14  3171.0 -8.00  50.33 -290.0 -20.0 -2.0   9.0  291.0
X15  3171.0 -7.78  50.07 -323.0 -18.0 -2.0   8.0  244.0
X16  3171.0 -7.38  49.82 -317.0 -14.0 -1.0   5.0  258.0
y    3171.0  0.51   0.50    0.0   0.0  1.0   1.0    1.0

In [11]:
#split explanatory variables from target variables
X_train = df_train.drop('y', axis=1) #explanatory
y_train = df_train['y'] #target

X_test = df_test.drop('y', axis=1)
y_test = df_test['y']


ecision Trees: Gini, Information Gain, Gain Ratio (6 marks)
Use any well-known library, Train three classifiers that differ only in the splitting criterion (Gini, Information Gain, Information Gain Ratio). Keep other settings default (set a fixed random_state for reproducibility). Evaluate each once on the provided test split and report binary accuracy.

In [12]:
from sklearn.tree import DecisionTreeClassifier
from C45 import C45Classifier

# https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html
# criterion{“gini”, “entropy”, “log_loss”}, default=”gini”
# The function to measure the quality of a split. Supported criteria are “gini” for the Gini impurity and “log_loss” and “entropy” both for the Shannon information gain, see Mathematical formulation.

# Train Decision Tree with Gini criterion
dt_gini = DecisionTreeClassifier(criterion='gini', random_state=7)
dt_gini.fit(X_train, y_train)

# Evaluate Gini-based model on the test set
accuracy_gini = dt_gini.score(X_test, y_test)
print(f"Accuracy with Gini criterion: {accuracy_gini:.5f}")

# Train Decision Tree with Information Gain criterion (entropy)
dt_inf_gain = DecisionTreeClassifier(criterion='entropy', random_state=7)
dt_inf_gain.fit(X_train, y_train)

# Evaluate the model
accuracy_inf_gain = dt_inf_gain.score(X_test, y_test)
print(f"Accuracy with Information Gain criterion: {accuracy_inf_gain:.5f}")

#Train Decision Tree with Information Gain Ratio criterion
dt_inf_gain_ratio = C45Classifier()
dt_inf_gain_ratio.fit(X_train, y_train)

#Evaluate the model
y_predict = pd.Series(dt_inf_gain_ratio.predict(X_test))  #predict returned list
total_matches = (y_test == y_predict).sum()
print(f"Accuracy with Information Gain Ratio criterion:\
 {total_matches / len(y_test):.5f}")


# accuracy_inf_gain_ratio = dt_inf_gain_ratio.evaluate(X_test, y_test)
# type(accuracy_inf_gain_ratio)
# print(f"Accuracy with Information Gain criterion: {accuracy_inf_gain_ratio:.5f}")


Accuracy with Gini criterion: 0.99517
Accuracy with Information Gain criterion: 0.99638
Accuracy with Information Gain Ratio criterion: 0.80941


The significantly lower accuracy of the Information Gain Ratio criterion (0.80941) compared to Gini (0.99517) and Information Gain (0.99638) is likely due to the Gain Ratio's inherent bias against attributes with a low number of distinct values.This outcome is a classic example of how a criterion designed to correct one bias can unintentionally introduce a new one, resulting in a suboptimal tree structure for certain datasets.Explanation of the Accuracy DropThe Purpose of Gain Ratio (and its Bias)The Information Gain Ratio was introduced by Ross Quinlan (for the C4.5 algorithm) to correct a major bias in Information Gain (IG).Information Gain (IG) Bias: IG favors attributes that have a large number of distinct values (e.g., a customer ID or a date/timestamp). Such attributes can partition the data into many small, pure subsets, resulting in a very high (and potentially misleading) IG score. This often leads to overfitting on the training data.Observation: Your IG model had the highest accuracy (0.99638), suggesting your data might contain a high-cardinality feature that IG heavily favored.Information Gain Ratio Correction: Gain Ratio attempts to fix this by normalizing Information Gain using a term called Split Information (which is the entropy of the attribute itself).$$\text{Gain Ratio} = \frac{\text{Information Gain}}{\text{Split Information}}$$How the Correction Backfires (The Cause of Low Accuracy)The denominator ($\text{Split Information}$) acts as a penalty:High Cardinality Feature: A feature with many distinct values (like an ID) has high Split Information, which heavily penalizes its Gain Ratio, reducing its chance of being selected. (This is good.)Low Cardinality Feature (The Problem): The same penalty mechanism causes the Gain Ratio to sometimes bias against attributes that create very uneven, unbalanced splits, even if those splits are highly informative. In extreme cases, if a split creates one very small partition and one very large partition, the Split Information can be near zero, causing the Gain Ratio to become unstable or incorrectly favor another, less informative split.Conclusion for Your DataBecause the Gain Ratio model's accuracy is so much lower, the most probable explanation is that:The Information Gain Ratio criterion consistently chose a suboptimal feature to split on at the root and subsequent nodes, likely because it penalized the truly best splitting feature(s) too heavily. This resulted in a less effective, poorly structured decision tree that could not capture the underlying patterns in the data as well as the Gini or pure IG models, leading to significant underfitting or an inefficient tree structure.The Key Takeaway: While IG Ratio corrects the bias toward high-cardinality features, it can be too aggressive, leading to an inferior tree when the most informative splits naturally lead to unbalanced partitions, as appears to be the case with your specific dataset.

https://gemini.google.com/app/37b7ef09558abc37

In [13]:
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

# Normalize the training and test data
scaler = StandardScaler() #z-scale normalization using std ev and mean
X_train_normalized = scaler.fit_transform(X_train) #derive mean and std dev from x tranin and mnormalize
X_test_normalized = scaler.transform(X_test) #normailize using std dev and mean derived from x_train

# Convert back to DataFrames for easier handling if needed later, keeping column names
X_train_normalized_df = pd.DataFrame(X_train_normalized, columns=X_train.columns)
X_test_normalized_df = pd.DataFrame(X_test_normalized, columns=X_test.columns)


# Define the same list of k values to test
k_values = [1, 3, 5, 7, 9, 15, 21, 49] # Using the same k values as before

# Dictionary to store accuracy for each k on normalized data
accuracy_scores_normalized = {}

print("Training and evaluating KNN on normalized data:")
for k in k_values:
    # Train KNN model with the current k on normalized data
    knn_normalized = KNeighborsClassifier(n_neighbors=k)
    knn_normalized.fit(X_train_normalized_df, y_train)

    # Predict on the normalized test set
    y_pred_knn_normalized = knn_normalized.predict(X_test_normalized_df)

    # Evaluate the model
    accuracy_normalized = accuracy_score(y_test, y_pred_knn_normalized)
    accuracy_scores_normalized[k] = accuracy_normalized
    print(f"Accuracy with KNN (k={k}) on normalized data: {accuracy_normalized:.4f}")

# Find the best k and its accuracy on normalized data
best_k_normalized = max(accuracy_scores_normalized, key=accuracy_scores_normalized.get)
best_accuracy_normalized = accuracy_scores_normalized[best_k_normalized]

print(f"\nBest k value on normalized data: {best_k_normalized} with accuracy: {best_accuracy_normalized:.4f}")

Training and evaluating KNN on normalized data:
Accuracy with KNN (k=1) on normalized data: 0.9964
Accuracy with KNN (k=3) on normalized data: 0.9952
Accuracy with KNN (k=5) on normalized data: 0.9916
Accuracy with KNN (k=7) on normalized data: 0.9891
Accuracy with KNN (k=9) on normalized data: 0.9879
Accuracy with KNN (k=15) on normalized data: 0.9807
Accuracy with KNN (k=21) on normalized data: 0.9795
Accuracy with KNN (k=49) on normalized data: 0.9578

Best k value on normalized data: 1 with accuracy: 0.9964


Random Forest baseline or K-Nearest Neighbour (KNN) (4 marks)
Train a single RandomForestClassifier (default settings; fixed random_state) or K-Nearest Neighbour (Change the value of k and see which is best for this problem) and evaluate once on the same test split. Report binary accuracy and add a one-sentence comment comparing it to the single-tree results.

In [14]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

# Define a list of k values to test
k_values = [1, 3, 5, 7, 9, 50, 100] # Testing 5 different k values

# Dictionary to store accuracy for each k
accuracy_scores = {}

for k in k_values:
    # Train KNN model with the current k
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train, y_train)

    # Predict on the test set
    y_pred_knn = knn.predict(X_test)

    # Evaluate the model
    accuracy = accuracy_score(y_test, y_pred_knn)
    accuracy_scores[k] = accuracy
    print(f"Accuracy with KNN (k={k}): {accuracy:.4f}")

# Find the best k and its accuracy
best_k = max(accuracy_scores, key=accuracy_scores.get)
best_accuracy = accuracy_scores[best_k]

print(f"\nBest k value: {best_k} with accuracy: {best_accuracy:.4f}")

Accuracy with KNN (k=1): 0.9964
Accuracy with KNN (k=3): 0.9952
Accuracy with KNN (k=5): 0.9916
Accuracy with KNN (k=7): 0.9891
Accuracy with KNN (k=9): 0.9879
Accuracy with KNN (k=50): 0.9590
Accuracy with KNN (k=100): 0.9276

Best k value: 1 with accuracy: 0.9964
